Speed up write Thermodynamic States

Lnaden opened the pull request:
This is an optimization to the `Reporter.write_thermodynamic_states` function to only serialize the `openmmtools.ThermodynamicState._standard_state` object if we have a new, incompatible `ThermodynamicState` with the states we already serialized. 

This results in about a 20% speedup to start time as a lower bound. The speed improvement is a function of both system size and number of states, although there does not appear to be a direct correlation as the `write_dict` and `serialize` functions become the rate-limiting steps.

This will also not pass tests until OpenMMTools 0.9.5 goes live

cc @andrrizzi for review

jchodera commented:
This doesn't require changing the reading of thermodynamic states too?

andrrizzi commented on Yank/repex.py:
```diff
@@ -401,39 +401,37 @@ def write_thermodynamic_states(self, thermodynamic_states, unsampled_states):
+                # If no compatible state is found, do full serialization
+                if not found_compatible_state:
+                    serialized_state = mmtools.utils.serialize(thermodynamic_state)
                     serialized_thermodynamic_state = serialized_state['thermodynamic_state']
```
I think this will raise `KeyError` for normal `ThermodynamicState`s (only `CompoundThermodynamicState`'s serialization has the `thermodynamic_state` key.

andrrizzi commented on Yank/repex.py:
```diff
@@ -401,39 +401,37 @@ def write_thermodynamic_states(self, thermodynamic_states, unsampled_states):
+                        serialized_thermodynamic_state = mmtools.utils.serialize(thermodynamic_state, skip_system=True)
+                        standard_system_hash = stored_states[compare_state]
+                        reference_state_name, len_serialization = compatible_states_hashes[standard_system_hash]
+                        serialized_thermodynamic_state['_Reporter__compatible_state'] = reference_state_name
```
I think you'll have to unnest the `ThermodynamicState` before adding this field or the current implementation of `read_thermodynamic_states` won't work.

Simply adding `_Reporter__compatible_state` should work, but you could also decide to pop the `standard_system` key from the thermodynamic state serialization to avoid serializing a bunch of `standard_system: None` in the dictionary. I don't have a strong preference here though.

andrrizzi commented on Yank/repex.py:
```diff
@@ -401,39 +401,37 @@ def write_thermodynamic_states(self, thermodynamic_states, unsampled_states):
-                # Write state as dictionary.
-                self.write_dict('{}/state{}'.format(state_type, state_id), serialized_state)
+                # Finally write the dictionary
+                self.write_dict('{}/state{}'.format(state_type, state_id), serialized_thermodynamic_state)
```
I think you want to serialize `serialized_state` here, not `serialized_thermodynamic_state`.

andrrizzi commented on devtools/conda-recipe/meta.yaml:
```diff
@@ -23,7 +23,7 @@ requirements:
     - ambermini >=16.16.0
     - docopt
-    - openmoltools
+    - openmoltools >=0.9.5
```
In the end we went for `0.10.0` because the release breaks the API.

andrrizzi commented:
Unfortunately tests here won't pass anyway because in `openmmtools-0.10.0` `AlchemicalFactory` has been renamed, and that's fixed in #685 . If `test_repex.py:TestReporter` pass locally it should be fine, but let me know if you want me to merge that branch in here instead that in `master`.

Lnaden commented:
@andrrizzi Okay, I made the changes requested. I switched out a few of the variable names to try and make things less confusing* so if you could double check this again to make sure I did not break anything new.

*i.e. `thermodynamic_state` from loop -> serialized to `serialized state` which may or may not have a "thermodynamic_state" dict key which is then extracted to `serialized_thermodynamic_state`. 

Changed this to `state` in the loop -> `serialized_state` of which the `serialized_thermodynamic_state` is extracted (and may be the same object, but thats okay)

Lnaden commented:
> This doesn't require changing the reading of thermodynamic states too?

@jchodera no, because this does not actually change what is written, just reduces the number of expensive calls to serialize. Instead of serializing every system to string, then figuring out what is is unique and dropping the string, it figures out what is unique first, then only serializes the system as needed.

andrrizzi commented:
The two problems I found were that `unnest_thermodynamic_state` was returning the wrong variable taken from the outer scope, and that the hack I implemented in `read_thermodynamic_states` didn't work anymore with the way things are handled in `openmmtools` now.

I've also removed the storing of `len_serialization` and the need to compute the hash at all in `write_thermodynamic_states` which should give a slight speed up. I've also modified `read_thermodynamic_states` so that we'll avoid serializing a system every time we use another state as reference, which should give us a great speed up during restore.

Tests pass locally now. @Lnaden could you double check the modifications before merging?
